File: kotlin/tax-system/src/main/kotlin/com/rbleggi/taxsystem/model/Product.kt
package com.rbleggi.taxsystem.model

sealed class Product(open val name: String, open val price: Double)

data class Electronic(override val name: String, override val price: Double) : Product(name, price)
data class Book(override val name: String, override val price: Double) : Product(name, price)
data class Food(override val name: String, override val price: Double) : Product(name, price)

File: scala-3.6/tax-system/src/main/scala/com/rbleggi/taxsystem/model/Product.scala
package com.rbleggi.taxsystem.model

sealed trait Product {
  def name: String
  def price: Double
}

final case class Electronic(name: String, price: Double) extends Product
final case class Book(name: String, price: Double) extends Product
final case class Food(name: String, price: Double) extends Product
